### PR TITLE
Stix 2.1 support

### PR DESCRIPTION
**Pull request by baulus**

### CODE REVIEW

**Review by ejratl — COMMENTED** [review]
In stix2matcher/matcher.py, lines 1062 & 1066, is there a case where the if statement on line 1062 would be true, but the if statement on line 1066 would be false? Is 1066 needed?

**Comment by baulus** [comment]
> In stix2matcher/matcher.py, lines 1062 & 1066, is there a case where the if statement on line 1062 would be true, but the if statement on line 1066 would be false? Is 1066 needed?

I don't see "if" statement on [line 1062](https://github.com/baulus/cti-pattern-matcher/blob/stix-2.1-support/stix2matcher/matcher.py#L1062) to compare. I guess you mean the "for" cycle. The data passed to the function (observed_data_sdos) is different in 2.0 and 2.1. In case of 2.1 we need to to additional operations to "number_observed" value.

**Comment by mdazam1942** [comment]
@clenk can you please have a look at this PR? Since this is an enhancement so we would like tor run by you before we merge it in?

File: README.rst
Comment (mdazam1942) [inline README.rst:71]: can you also update the usage command with this new parameter? 
`usage: stix2-matcher [-h] -p PATTERNS -f FILE [-e ENCODING] [-v]`

Comment (baulus) [reply README.rst]: thx, done

File: stix2matcher/matcher.py
Comment (baulus) [reply stix2matcher/matcher.py]: Added DEFAULT_VERSION

Comment (baulus) [reply stix2matcher/matcher.py]: Are we good to merge?

**Comment by baulus** [comment]
I am not quite sure why the check does not pass https://github.com/oasis-open/cti-pattern-matcher/runs/6779347901?check_suite_focus=true#step:5:334

**Comment by clenk** [comment]
> I am not quite sure why the check does not pass https://github.com/oasis-open/cti-pattern-matcher/runs/6779347901?check_suite_focus=true#step:5:334 what import sorting is required? Any ideas @clenk ?

It shows the change needed in the output: https://github.com/oasis-open/cti-pattern-matcher/runs/6779347901?check_suite_focus=true#step:5:348 (the lines that start with `-` and `+`)

**Comment by chisholm** [comment]
Here are some of my observations.  The matcher seems to exhibit some odd behavior.

Expected input seems to now be a list of bundles.  Consider the following STIX content to be matched against (a list with one bundle; this content is a derivative of one of the unit tests):

```json
[
    {
        "id": "bundle--c2a23389-17c9-4247-8aa2-d2ffde3de334",
        "type": "bundle",
        "objects": [
            {
                "id": "observed-data--647947aa-4cc3-43a3-a42f-2c670986d777",
                "type": "observed-data",
                "spec_version": "2.1",
                "number_observed": 1,
                "first_observed": "2014-04-19T06:51:26Z",
                "last_observed": "2014-04-19T06:51:26Z",
                "object_refs": [
                    "test--9c61d8ff-21ca-489a-accd-7356c2ea0a7b"
                ]
            },
            {
                "id": "test--9c61d8ff-21ca-489a-accd-7356c2ea0a7b",
                "type": "test",
                "spec_version": "2.1",
                "int": 5
            },
            {
                "id": "test2--ac61d8ff-21ca-489a-accd-7356c2ea0a7b",
                "type": "test2",
                "spec_version": "2.1",
                "int": 5
            }
        ]
    }
]
```

I observe the following:

- The matcher seems to match against anything in a bundle, regardless of STIX type or whether it is part of an observation.  You can even match the observation itself as if it were an SCO.  For example,

        [observed-data:number_observed = 1]

    matches.  This should not match, because comparison expressions only match SCOs which are part of an observation.  observed-data objects are not SCOs.

- It will match SCOs which are not part of any observation.  For example,

        [test2:int = 5]

    matches.  This should not match, because the test2 object is not a part of any observation.

These matches can be confirmed with the commandline tool, e.g.

    stix2-matcher -s 2.1 -f .\bundles.json -p .\patterns.txt

(where the above content is in bundles.json and the patterns are in patterns.txt.)

Additionally, the `match()` function now seems to return lists of bundles.  In order to find matches to STIX patterns, the matcher must find bindings of observations to observation expressions.  To help users understand the matches they get, I think it ought to return those bindings, which means returning the observations.

A bundle need not contain only one observation, right?  Are there assumptions about what each bundle must contain?  If so, that should be documented, and maybe checked to ensure users don't get garbage results due to malformed bundles.  Better yet, I think the matcher should be flexible in what it accepts.  I think it should be easy to include the observations and those SCOs they refer to via object_refs, and ignore everything else.  Bundles are fine as a means of content delivery, but since they have no meaning, I think the matcher should internally act as if they did not exist (e.g. discard them and only use their contents).

**Comment by baulus** [comment]
Thanks @chisholm  for your reviewing and bringing the examples.
Can I ask to bring example of json (whatever part of bundle it be) as input and the output of `match()` function, for version 2.1. 
Is the input you think is just the content of the "object" field of the bundle (list of the lists, considering there can be multiple bundles)? And the output is the same list of lists without "observed-data" elements, and without the elements that did not match or are not referenced in matched elements?
Thanks in advance.

**Comment by chisholm** [comment]
It sounds like you are asking about the structure of input (as opposed to its content)?  As mentioned above, I think the structure could be flexible, as long as it contains the minimum necessary content to allow the matcher to do its job.  The minimum necessary content includes the observations and all SCOs they reference.

I think the simplest possible input structure is probably an ordinary list or iterable of STIX objects, similar to the input for the old matcher.

For example:

```json
[
    {
        "type": "observed-data",
        "spec_version": "2.1",
        "id": "observed-data--1209f166-bba6-4307-a566-60e3a6c39cd2",
        "created": "2022-10-25T08:59:09.295455Z",
        "modified": "2022-10-29T01:53:18.132077Z",
        "first_observed": "2022-11-27T19:50:46.889315Z",
        "last_observed": "2023-01-05T05:35:31.372958Z",
        "number_observed": 1,
        "object_refs": [
            "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
        ]
    },
    {
        "type": "observed-data",
        "spec_version": "2.1",
        "id": "observed-data--9813cee6-9e9d-4199-88d2-a740331fa8a1",
        "created": "2021-09-17T05:06:25.871976Z",
        "modified": "2022-06-19T20:23:09.126157Z",
        "first_observed": "2022-11-05T10:47:16.435447Z",
        "last_observed": "2023-01-17T23:00:47.872163Z",
        "number_observed": 1,
        "object_refs": [
            "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
        ]
    },
    {
        "type": "ipv4-addr",
        "spec_version": "2.1",
        "id": "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5",
        "value": "55.73.253.84"
    }
]
```

The above includes two observations, each of which shares the same SCO.  The return value of the matcher was intended to be explanatory: to help users understand how and why a pattern matched.  So I think output ought to include the observations which bind to the observation expressions and produce the match.  For example, if the pattern were:

    [ipv4-addr:value = '55.73.253.84'] REPEATS 2 TIMES

Then minimal output would be:

```json
[
    {
        "type": "observed-data",
        "spec_version": "2.1",
        "id": "observed-data--1209f166-bba6-4307-a566-60e3a6c39cd2",
        "created": "2022-10-25T08:59:09.295455Z",
        "modified": "2022-10-29T01:53:18.132077Z",
        "first_observed": "2022-11-27T19:50:46.889315Z",
        "last_observed": "2023-01-05T05:35:31.372958Z",
        "number_observed": 1,
        "object_refs": [
            "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
        ]
    },
    {
        "type": "observed-data",
        "spec_version": "2.1",
        "id": "observed-data--9813cee6-9e9d-4199-88d2-a740331fa8a1",
        "created": "2021-09-17T05:06:25.871976Z",
        "modified": "2022-06-19T20:23:09.126157Z",
        "first_observed": "2022-11-05T10:47:16.435447Z",
        "last_observed": "2023-01-17T23:00:47.872163Z",
        "number_observed": 1,
        "object_refs": [
            "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
        ]
    }
]
```

I would say it is not necessary to return the referenced SCOs as well.

So your characterization of the output as "output is the same list of lists without "observed-data" elements..." is incorrect I think.  I think the bound observed-data objects are the *most* important things to return.

In the spirit of flexibility, here are some additional input structures.  Output should be the same as long as the content of the input is the same, regardless of how that input is structured.

The following is a list-of-lists structure, where there is one observation and related SCOs per inner list.  This may have been closer to what you were originally thinking?

```json
[
    [
        {
            "type": "observed-data",
            "spec_version": "2.1",
            "id": "observed-data--1209f166-bba6-4307-a566-60e3a6c39cd2",
            "created": "2022-10-25T08:59:09.295455Z",
            "modified": "2022-10-29T01:53:18.132077Z",
            "first_observed": "2022-11-27T19:50:46.889315Z",
            "last_observed": "2023-01-05T05:35:31.372958Z",
            "number_observed": 1,
            "object_refs": [
                "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
            ]
        },
        {
            "type": "ipv4-addr",
            "spec_version": "2.1",
            "id": "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5",
            "value": "55.73.253.84"
        }
    ],
    
    [
        {
            "type": "observed-data",
            "spec_version": "2.1",
            "id": "observed-data--9813cee6-9e9d-4199-88d2-a740331fa8a1",
            "created": "2021-09-17T05:06:25.871976Z",
            "modified": "2022-06-19T20:23:09.126157Z",
            "first_observed": "2022-11-05T10:47:16.435447Z",
            "last_observed": "2023-01-17T23:00:47.872163Z",
            "number_observed": 1,
            "object_refs": [
                "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
            ]
        },
        {
            "type": "ipv4-addr",
            "spec_version": "2.1",
            "id": "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5",
            "value": "55.73.253.84"
        }
    ]
]
```

With this type of structure the SCO is repeated, making it more verbose.

The above two structures can use bundles, for additional variations (the following leave out some required bundle properties for brevity):

```json
{
    "type": "bundle",
    "objects": [
        {
            "type": "observed-data",
            "spec_version": "2.1",
            "id": "observed-data--1209f166-bba6-4307-a566-60e3a6c39cd2",
            "created": "2022-10-25T08:59:09.295455Z",
            "modified": "2022-10-29T01:53:18.132077Z",
            "first_observed": "2022-11-27T19:50:46.889315Z",
            "last_observed": "2023-01-05T05:35:31.372958Z",
            "number_observed": 1,
            "object_refs": [
                "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
            ]
        },
        {
            "type": "observed-data",
            "spec_version": "2.1",
            "id": "observed-data--9813cee6-9e9d-4199-88d2-a740331fa8a1",
            "created": "2021-09-17T05:06:25.871976Z",
            "modified": "2022-06-19T20:23:09.126157Z",
            "first_observed": "2022-11-05T10:47:16.435447Z",
            "last_observed": "2023-01-17T23:00:47.872163Z",
            "number_observed": 1,
            "object_refs": [
                "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
            ]
        },
        {
            "type": "ipv4-addr",
            "spec_version": "2.1",
            "id": "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5",
            "value": "55.73.253.84"
        }
    ]
}
```

and

```json
[
    {
        "type": "bundle",
        "objects": [
            {
                "type": "observed-data",
                "spec_version": "2.1",
                "id": "observed-data--1209f166-bba6-4307-a566-60e3a6c39cd2",
                "created": "2022-10-25T08:59:09.295455Z",
                "modified": "2022-10-29T01:53:18.132077Z",
                "first_observed": "2022-11-27T19:50:46.889315Z",
                "last_observed": "2023-01-05T05:35:31.372958Z",
                "number_observed": 1,
                "object_refs": [
                    "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
                ]
            },
            {
                "type": "ipv4-addr",
                "spec_version": "2.1",
                "id": "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5",
                "value": "55.73.253.84"
            }
        ]
    },
    {
        "type": "bundle",
        "objects": [
            {
                "type": "observed-data",
                "spec_version": "2.1",
                "id": "observed-data--9813cee6-9e9d-4199-88d2-a740331fa8a1",
                "created": "2021-09-17T05:06:25.871976Z",
                "modified": "2022-06-19T20:23:09.126157Z",
                "first_observed": "2022-11-05T10:47:16.435447Z",
                "last_observed": "2023-01-17T23:00:47.872163Z",
                "number_observed": 1,
                "object_refs": [
                    "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5"
                ]
            },
            {
                "type": "ipv4-addr",
                "spec_version": "2.1",
                "id": "ipv4-addr--0ddfa859-1a40-4aa0-b440-136c498cc5a5",
                "value": "55.73.253.84"
            }
        ]
    }
]
```

Again, input structure can be flexible, but should include whatever you think the most likely structure is that users will have, so there is minimal burden imposed on them to transform their content into what the matcher expects.  If it were me, I think I would support the simple flat list/iterable input structure at minimum.

**Comment by delliott90** [comment]
I think the second to last example is what we would like. The observed-data objects are returned along with the unique SCOs that they reference. That is how stix-shifter is returning STIX 2.1 results, and that's based on the examples given in the 2.1 spec.

**Comment by chisholm** [comment]
It looks much better.  Looks like the input requirement is more flexible now but still requires bundles, and requires that an observation and its referenced SCOs can't be split across multiple bundles.  They all have to be together.

> Just one observation that I found while implementing this: based on your first example in your last comment, if we have 2 matches for `[ipv4-addr:value = '55.73.253.84']`, the matcher will only return first occurrence of the sdo and its referenced cso objects matching the patter. The second `observed-data--9813cee6-9e9d-4199-88d2-a740331fa8a1` will not be included in the result. It is [filtered out here](https://github.com/oasis-open/cti-pattern-matcher/blob/4ee507e2df43a3c491db4c841512f04792bc867a/stix2matcher/matcher.py#L2368) so the result will look like:

It was intentional to return only the first binding (it was not clear that multiple explanations for a match would be useful; you just needed *some* explanation to show that the pattern did match).  So you just need to ensure that a complete explanation is returned, e.g. for a `REPEATS 2 TIMES` pattern that you return a size-two binding (at least), since that is what the pattern requires.  And it looks like that is still the case, so that's good.

This funny special case (with respect to `type_`) caught my eye:

https://github.com/baulus/cti-pattern-matcher/blob/55f7c76512f21a188a4ab3b54e22235158e44f7a/stix2matcher/matcher.py#L2210

So there is hard-coding to skip observations and identities as object path types.  Really, you'd want to skip any non-SCO.  But most non-SCOs shouldn't appear in an observation anyway.  There actually *is* one non-SCO which is legal in a STIX 2.1 observation, and that's a relationship SRO.  So I came up with the following example:

```json
[
    {
        "id": "bundle--c2a23389-17c9-4247-8aa2-d2ffde3de334",
        "type": "bundle",
        "objects": [
            {
                "id": "observed-data--647947aa-4cc3-43a3-a42f-2c670986d777",
                "type": "observed-data",
                "spec_version": "2.1",
                "number_observed": 1,
                "first_observed": "2014-04-19T06:51:26Z",
                "last_observed": "2014-04-19T06:51:26Z",
                "object_refs": [
                    "ipv4-addr--9c61d8ff-21ca-489a-accd-7356c2ea0a7b",
                    "domain-name--57196be9-5512-4867-8c86-8c622c741c31",
                    "relationship--e9a303d2-3b43-4e20-aedd-29cd9afede78"
                ]
            },
            {
                "id": "ipv4-addr--9c61d8ff-21ca-489a-accd-7356c2ea0a7b",
                "type": "ipv4-addr",
                "spec_version": "2.1",
                "value": "1.2.3.4"
            },
            {
                "id": "domain-name--57196be9-5512-4867-8c86-8c622c741c31",
                "type": "domain-name",
                "spec_version": "2.1",
                "value": "example.com"
            },
            {
                "id": "relationship--e9a303d2-3b43-4e20-aedd-29cd9afede78",
                "type": "relationship",
                "spec_version": "2.1",
                "relationship_type": "related-to",
                "source_ref": "ipv4-addr--9c61d8ff-21ca-489a-accd-7356c2ea0a7b",
                "target_ref": "domain-name--57196be9-5512-4867-8c86-8c622c741c31"
            }
        ]
    }
]
```

and pattern:

    [relationship:relationship_type = 'related-to']

This content is STIX compliant and actually results in a match, but it shouldn't because relationships are not SCOs.

**Comment by baulus** [comment]
hi @chisholm the recommendations are implemented, please take a look, basically I learn new things about the stix spec as I implement them. 
Thanks for the explanation for REPEATS matching results.

For searching in multiple "bundles", at this moment I saw the most efficient to merge multiple bundles into one, once, at the very beginning, and then let the script do the regular matching, rather than modify the script even more to have it searching in all bundles (that might add more complexity).  

For the non-SCO types, a list of stopper type names are added, retrieved from [cti-stix-validator project ](https://github.com/oasis-open/cti-stix-validator/blob/1e35d84e4d970d61c16201d217eaabfebbd5e932/stix2validator/v21/enums.py#L412). I thought it is not a good idea to include cti-stix-validator module as required only for this purpose. 

Please, let me know if the changes are efficient and cover your recommendations.

**Comment by chisholm** [comment]
Yes, I see they are merged, and then re-split into new bundles.  Looks like that issue is now fixed.

Here is a question: according to STIX 2.1 spec section 12.7.1, STIX patterning conformance level 1, which is the minimum level, requires support for all comparison operators.  Are there plans to add support for the `EXISTS` comparison operator, which was new in 2.1?

**Comment by delliott90** [comment]
> Here is a question: according to STIX 2.1 spec section 12.7.1, STIX patterning conformance level 1, which is the minimum level, requires support for all comparison operators. Are there plans to add support for the `EXISTS` comparison operator, which was new in 2.1?

It would be a good idea to support it but I'd be inclined to open a new issue to add it rather than further block this PR.

**Review by mdazam1942 — APPROVED** [review]

**Pull request merged**
